### PR TITLE
Change tag content to JSON

### PR DESCRIPTION
**Pull request by TomGoBravo**
This PR builds towards https://trello.com/c/Cq2zNk2M/674-annotate-data-outliers-that-weve-flagged-in-the-ui

* Moves `date` from a field in the tag/annotation CSV and DataFrame to a field in the JSON content of annotations. This replaces the human readable strings buried in our data pipeline with flexible JSON content. The downside is processing the JSON in pandas is harder. The upside is this makes adding new attributes per annotation *much* easier. Removing the date will also make it easier to include all the annotations in the wide-dates CSV so we can later remove multiregion-annotations.csv.

### CODE REVIEW

File: libs/datasets/timeseries.py [hunk 1]
```diff
@@ -82,20 +76,146 @@ class TagType(GetByValueMixin, ValueAsStrMixin, str, enum.Enum):
...
+TAG_TYPE_TO_CLASS = {
+    TagType.CUMULATIVE_TAIL_TRUNCATED: CumulativeTailTruncated,
+    TagType.CUMULATIVE_LONG_TAIL_TRUNCATED: CumulativeLongTailTruncated,
+    TagType.ZSCORE_OUTLIER: ZScoreOutlier,
+    TagType.PROVENANCE: ProvenanceTag,
+}
+
+
+@dataclass(frozen=True)
+class TagCollection:
```
Comment (ghop02) [inline libs/datasets/timeseries.py:197]: Thoughts on throwing this in another module?

File: test/test_helpers.py [hunk 2]
```diff
@@ -58,8 +56,28 @@ def make_tag(
     type: TagType = TagType.CUMULATIVE_TAIL_TRUNCATED,
     date: Union[pd.Timestamp, str] = "2020-04-02",
     content: str = "taggy",
+    **kwargs,
 ) -> timeseries.TagInTimeseries:
-    return timeseries.TagInTimeseries(type, pd.to_datetime(date), content)
+    date: pd.Timestamp = pd.to_datetime(date)
+    if type == TagType.PROVENANCE:
+        assert content != "taggy"
```
Comment (ghop02) [inline test/test_helpers.py:63]: What are these checks for?

File: libs/datasets/timeseries.py [hunk 3]
```diff
@@ -82,20 +76,146 @@ class TagType(GetByValueMixin, ValueAsStrMixin, str, enum.Enum):
...
+
+    _location_var_map: MutableMapping[Tuple, List[TagInTimeseries]] = dataclasses.field(
+        default_factory=lambda: collections.defaultdict(list)
+    )
+
+    def add(self, tag: TagInTimeseries, *, location_id: str, variable: CommonFields) -> None:
+        """Adds a tag to this collection."""
+        self._location_var_map[(location_id, variable)].append(tag)
+
+    def _as_records(self) -> Iterable[Mapping]:
```
Comment (ghop02) [inline libs/datasets/timeseries.py:209]: maybe
```
@property
def _records(self)
```
and 
```
@property
def dataframe(self)
```
?

but don't feel strongly
```

**Review by ghop02 — APPROVED** [review]
This looks good to me!

File: test/test_helpers.py (hunk 2 above)
Comment (TomGoBravo) [reply test/test_helpers.py]: They were leftover from a transition from a previous version of the code. removed.

**Pull request merged**